src/components/KeyboardAvoidingView/hooks.ts

```diff
@@ -1,14 +1,20 @@
+import { useState } from "react";
 import { useSharedValue } from "react-native-reanimated";
 
 import { useKeyboardContext } from "../../context";
 import { useKeyboardHandler } from "../../hooks";
 
 export const useKeyboardAnimation = () => {
   const { reanimated } = useKeyboardContext();
-  const heightWhenOpened = useSharedValue(-reanimated.height.value);
-  const height = useSharedValue(-reanimated.height.value);
-  const progress = useSharedValue(reanimated.progress.value);
-  const isClosed = useSharedValue(reanimated.progress.value === 0);
+
+  // calculate it only once on mount, to avoid `SharedValue` reads during a render
+  const [initialHeight] = useState(() => -reanimated.height.value);
+  const [initialProgress] = useState(() => reanimated.progress.value);
```

Comment:
hm, they already use useState under the hood https://github.com/software-mansion/react-native-reanimated/blob/main/packages/react-native-reanimated/src/hook/useSharedValue.ts#L19 and if we have a look at the react docs 

initialState: The value you want the state to be initially. It can be a value of any type, but there is a special behavior for functions. This argument is ignored after the initial render.

So, are you sure that you need these changes?

Reply:
Yeah, but new REA 3.16 will complain if you use `.value` during render. So you should void the usage of `.value` during re-render (but you can use for the first render).

I had several approaches:

1. To use `useMemo` to calculate initial values for `reanimated.*.value` - but this is not reliable, because `useMemo` in future may throw cache away and re-calculate memoization.

2. To use `useState` to calculate initial value.

3. to create own `useLazySharedValue` hook with next implementation:

```ts
export function useSharedValue<Value>(initialValue: () => Value): SharedValue<Value> { // function instead of value
  const [mutable] = useState(() => makeMutable(initialValue()));
  useEffect(() => {
    return () => {
      cancelAnimation(mutable);
    };
  }, [mutable]);
  return mutable;
}
```

If you have more ideas @IvanIhnatsiuk then I'll be happy to consider your thoughts 😊

Reply:
hm, looks like it would be better to implement your own hook and remove it once people migrate to REA 3.16

Also, it helps to remove code duplication like this:

```.tsx
 const [initialHeight] = useState(() => -reanimated.height.value);
 const heightWhenOpened = useSharedValue(initialHeight);
```

and replace it with your own hook

```.tsx
 const heightWhenOpened = useLazySharedValue(() => -reanimated.height.value);
```

But anyway, it's up to you 🙂

Reply:
Well, there are few reasons why I don't want to implement my hook:
- `makeMutable` is publicly exported, but I don't want to copy internal implementation of `useSharedValue` (even though this hook is a very simple). If it changes over time I'll have to replicate these changes in this package as well.
- we have kind of code duplication, but thus we extract values and then reuse the value (I'm not sure if reading `SharedValue` on first render is an expensive operation or not, but I saw somewhere that it adds an overhead - but maybe mistaken).

This is why I think `useState` approach is better. anyway, it's pretty minor code modification - in future we may revisit this code 🙂